Risk Score Persistence API

rylnd opened the pull request:
## Summary

*  Introduces a new API, POST `/api/risk_scores/calculate`, that triggers the code introduced here
    * As with the [preview route](https://github.com/elastic/kibana/pull/155966), this endpoint is behind the `riskScoringRoutesEnabled` feature flag
    * We intend to __REMOVE__ this endpoint before 8.10 release; it's mainly a convenience/checkpoint for testing the existing code. The next PR will introduce a scheduled Task Manager task that invokes this code periodically.
* Updates to the /preview route:
  * `data_view_id` is now a required parameter on both endpoints. If a dataview is not found by that ID, the id is used as the general index pattern to the query.
  * Response has been updated to be more similar to the [ECS risk fields](https://github.com/elastic/ecs/pull/2236) powering this data.
* Mappings created by the [Data Client](https://github.com/elastic/kibana/pull/158422) have been updated to be aligned to the ECS risk fields (linked above)
* Adds/updates the [OpenAPI spec](https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/server/lib/risk_engine/schema/risk_score_apis.yml) for these endpoints; useful starting point if you're trying to get oriented here.


## Things to review
* [PR Demo environment](https://rylnd-pr-161503-risk-score-task-api.kbndev.co/app/home)
* Preview API and related UI still works as expected
* Calculation/Persistence API correctly bootstraps/persists data
    * correct mappings/ILM are created
    * things work in non-default spaces
   



### Checklist


- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [ ] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios


### Risk Matrix

Delete this section if it is not applicable to this PR.

Before closing this PR, invite QA, stakeholders, and other developers to identify risks that should be tested prior to the change/feature release.

When forming the risk matrix, consider some of the following examples and how they may potentially impact the change:

| Risk                      | Probability | Severity | Mitigation/Notes        |
|---------------------------|-------------|----------|-------------------------|
| Multiple Spaces&mdash;unexpected behavior in non-default Kibana Space. | Low | High | Integration tests will verify that all features are still supported in non-default Kibana Space and when user switches between spaces. |
| Multiple nodes&mdash;Elasticsearch polling might have race conditions when multiple Kibana nodes are polling for the same tasks. | High | Low | Tasks are idempotent, so executing them multiple times will not result in logical error, but will degrade performance. To test for this case we add plenty of unit tests around this logic and document manual testing procedure. |
| Code should gracefully handle cases when feature X or plugin Y are disabled. | Medium | High | Unit tests will verify that any feature flag or plugin combination still results in our service operational. |
| [See more potential risk examples](https://github.com/elastic/kibana/blob/main/RISK_MATRIX.mdx) |


### For maintainers

- [ ] This was checked for breaking API changes and was [labeled appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)

rylnd committed:
Rename getScores -> calculateScores

Somewhat less ambiguous!

rylnd committed:
WIP: Adding new public endpoint and service methods

This defines most of the endpoint, along with unit tests, and the
skeleton of the service methods.

There are a few type errors to pick up next, as we still need to define
the return type of the service method.

rylnd committed:
Rename integration test file to be more accurate

rylnd committed:
Placeholder integration test file

rylnd committed:
Rename factory function as such

rylnd committed:
Fleshes out more of the persistence implementation

* More realistic parameters, types
* Threads through the data client
* Calls the unimplemented bulk method on the writer

It's a bit silly to have to pass the spaceId to the writer; I'd like
that to be a convenience method on the request context since it already
has that information.

rylnd committed:
Support dataview runtime fields in Risk Score calculations

data_view_id is now a required parameter on the preview route as well as
the new route. If a dataview is not found by that ID, it will be used
instead as an index pattern on the request. If a dataview is found, its
runtime fields will be applied to the risk score calculation request.

rylnd committed:
WIP: Implementing the actual persistence

* Registers the persistence route (temporarily)
* Adds a basic integration test exercising the route, and asserting docs
  were written
* Moves some risk_engine tests around, adds/abstracts some risk-engine-specific
  utils
* Writer should probably be extracted to a class (or at least a module)
  with its own tests.

rylnd committed:
Types/test cleanup

Following the previous commit, this fixes some typings, mocks, and
updates some now-invalid unit tests (since we changed the behavior of
data_view_id in the preview route.

rylnd committed:
Fix Risk Score Preview integration tests

We now require a data_view_id to be specified.

rylnd committed:
Proper cleanup of risk score preview integration tests

We should delete scores after each test so as not to pollute the
environment for the next one.

rylnd committed:
Wait for risk scores to be available before making assertions

This test was failing, and eventually it occurred to me why: we're not
using `wait_for` when writing risk scores, so depending on the index
refresh interval scores may not be immediately available.

This adds and uses a utility to address this. As for the implications
for this and serverless, I will discuss with product.

rylnd committed:
Fix moved schema definition

rylnd committed:
Extract Writer implementation to a separate class

rylnd committed:
Remove unused translations

We now require a data view in all our endpoints, and if one is not found
we fall back to using the dataview name/ID as the index pattern.

rylnd committed:
Fix issues following rebase onto EA management UI work

* Fix type import that was renamed
* Adds required `data_view_id` parameter to preview endpoint (needs
  verification that logic is correct)

rylnd committed:
Generate ECS mappings for risk fields

These mappings were previously based on what the API is returning, which
was not correct.

rylnd committed:
Update Risk Engine to handle ECS risk fields

* Updates APIs to return host and user scores under host and user keys,
  respectively
    * However, we trim out the unnecessary '.host.risk' keys from the
      individual scores, as those are redundant on the API (and would
      make a lot of unnecessary client work besides).
* Updates types and tests accordingly.

rylnd committed:
Fix missed test failures

Some were related to type issues that I didn't catch locally, others
were related to an untyped API stub that didn't catch this change
(cypress).

rylnd committed:
Fix Risk Score Preview table following API change

The columns being passed to EuiInMemoryTable are correctly inferred to
be `EuiBasicTableColumn<RiskScore>`, but unfortunately that type does not
restrict the field name to be `keyof RiskScore`. I suppose that this is
to allow the use of runtime fields, etc.

In our case, I'd like this to be a little more restrictive, so I made
the local type thus.

rylnd committed:
Update preview API doc to indicate `data_view_id` as a required field

rylnd committed:
Update OpenAPI spec for preview endpoint

* Updates fields to more closely align with ECS risk fields
* Updates examples (particularly category-related fields) to have more
  realistic examples

rylnd committed:
Adds OpenAPI schema for new (temporary) /calculate endpoint

rylnd committed:
Add unit tests for our DataWriter class

rylnd committed:
Adds a few more tests for mixed response cases

rylnd committed:
Merge branch 'main' into risk_score_task_api

elasticmachine commented:
Pinging @elastic/security-solution (Team: SecuritySolution)

nkhristinin reviewed (approved):
LGTM!

nkhristinin commented on x-pack/plugins/security_solution/server/lib/risk_engine/risk_weights.ts:
```diff
@@ -28,7 +28,7 @@ const DEFAULT_CATEGORY_WEIGHTS: RiskWeights = RISK_CATEGORIES.map((category) =>
  */
 const convertCategoryToEventKindValue = (category?: string): string | undefined =>
-  category === 'alerts' ? 'signal' : category;
+  category === 'category_1' ? 'signal' : category;
```
Why do we need to rename alerts to `cateogry_1`

rylnd replied:
This is discussed in the [ongoing ECS RFC(s)[https://github.com/elastic/ecs/pull/2232/files#diff-40e16fa32db2ffbdeabf56eb2ffeac75dfa12410e662c3328c7c859b426805d2R20], but the summary is: futureproofing.

nkhristinin commented:
Hey, I discover one confusing behaviour/bug

So, let's say we have 1500 alerts

1. call calculation and persist API - get results like

```
{
    "after_keys": {
        "host": {
            "host.name": "Host 548"
        }
    },
    "errors": [],
    "scores_written": 1000
}
```

2. Call it again with 
```
"after_keys": {
        "host": {
            "host.name": "Host 548"
        }
    }
```
    
And get results like:
   
```
{
    "after_keys": {
        "host": {
            "host.name": "Host 999"
        }
    },
    "errors": [],
    "scores_written": 500
}
```

So `scores_written: 500`, is less than 1000, which means we scored everything we can now.

3. But if I call it again with after_keys from step 2 
```
"after_keys": {
        "host": {
            "host.name": "Host 999"
        }
    }
 ```
 it will return status 200, but response:
    
 ```
    {
    "after_keys": {},
    "errors": [
        "parse_exception\n\tRoot causes:\n\t\tparse_exception: request body is required"
    ],
    "scores_written": 0
}
```

nkhristinin commented:
Would it be possible to add `host.name` and `user.name` fields to risks?

rylnd committed:
Add host.name and user.name to risk index mappings

Also renames the mapping declaration to be more accurate; it's not just
"risk fields" in here.

rylnd committed:
Persist host.name and user.name to risk score docs

Updates tests accordingly.

rylnd committed:
Less noisy integration test logs

We can log these when debugging a particular test; this shouldn't be the
default.

rylnd committed:
Short-circuit Risk Score persistence when there's nothing to persist

Also returns a sentinel value (documented in the OpenAPI schemae)
indicating this state.

rylnd commented:
> Hey, I discover one confusing behaviour/bug

Good catch, I addressed this in [94d0903](https://github.com/elastic/kibana/pull/161503/commits/94d090396b1eaab4d5f506c6f2089ff9f421e577). Let me know what you think!

rylnd committed:
Merge branch 'main' into risk_score_task_api

rylnd committed:
Update risk score fields following feedback from ECS team

See: https://github.com/elastic/ecs/pull/2236/commits/2372eb7c9e709e10140ad1932b590f5019b1d301

rylnd committed:
Merge branch 'main' into risk_score_task_api

rylnd committed:
Update snapshot tests

I had changed these values manually, but the order changed because:
alphabet.

rylnd committed:
Merge branch 'main' into risk_score_task_api

nkhristinin commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into risk_score_task_api
